policymap, fragmap: clean up doc, map size configuration

qmonnet opened the pull request:
More clean-up following the introduction of IPv4 fragment tracking.

Follow-up on #10927.

First commit updates the doc for kubeproxy-free (removes note about missing fragmentation support).

Second commit cleans up the way we configure map size for policymap and fragmap, and falls back to limit values if the user passes unreasonable values. No lower limit for fragmap (other than 0), maybe we could remove it for policymap as well but I'm unsure if the lower limit is used elsewhere as an assumption. Anyway it's already possible to ask for as few as 256 entries in policy map, not sure it's worth trying to allow fewer than that.

tklauser commented on pkg/defaults/defaults.go:
```diff
@@ -246,6 +246,14 @@ const (
 	// the endpoint could send/receive traffic to/from.
 	PolicyMapEntries = 16384 // Cilium 1.5 and earlier value
 
+	// PolicyMapMinSize is the minimum value to which the number of entries
```
This will conflict with the refactoring in #10957 (for #10780) where these consts are moved to the top of `pkg/option/config` (like the other map min/max constants). Same for `FragmentMapMaxSize`. I think we can keep them in `pkg/option` as they are not used outside of the package.

I think it makes sense to keep the default map sizes in `pkg/defaults` though.

qmonnet replied:
With your change they would be in a list of similar limits for the other maps, which looks better than leaving them in the middle of the `config` package as it is now, so OK.

I'll probably hold until your PR is merged in that case, and rebase afterwards.

Thanks!

pchaigno reviewed (approved):
LGTM!

tklauser reviewed (approved):
Thanks, LGTM!

pchaigno reviewed (changes requested):
Actually, could you also fix up [the sentences in `config.go`](https://github.com/cilium/cilium/blob/1000f7023a46f2ff0c29b3b155c294942b5c4fd7/pkg/option/config.go#L2752)? That succession of compound adjectives without any punctuation is kinda hard to parse.

qmonnet committed:
fragmap, policymap: clean-up map size configuration

As a follow-up to IPv4 fragmentation cleanup (#10927), clean-up
configuration for the maximum number of entries in policy map and
fragment-tracking map.

Remove unnecessary InitMapInfo() for fragmap. Keep it for policy map,
because passing the size at map creation and removing this function
would mean:

- Updating the list of arguments passed to Open(), OpenAndCreate(),
  Create() and pass the size (option.Config.PolicyMapEntries) each time,
  which seems verbose and not really useful. We cannot import the
  "option" package in "policy" to use the option just once in newMap().

- Making sure that policymap.MaxEntries is never called from other
  packages before the map is created. This is probably not the case so
  we do want to set the size before map creation.

Initialise policymap.MaxEntries to defaults.PolicyMapEntries.
Theoretically we don't have to initialise it because it is overwritten
by InitMapInfo() before use, but the unit tests do not call
InitMapInfo() and rely a non-zero value to successfully create a map
(eppolicymap unit tests also use the value but does not care about map
creation being successful).

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
docs: update doc on limits for kubeproxy-free

Now that support for IPv4 fragments has been merged, remove the note in
kubeproxy-free documentation telling that fragments are dropped.

Reported-by: @pchaigno
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno reviewed (approved):
Nice catch on the null dereference for eppolicymap!
That will require a backport to 1.7, no?

tgraf commented on pkg/maps/eppolicymap/eppolicymap.go:
```diff
@@ -60,6 +60,18 @@ var (
 // for testing purposes.
 func CreateWithName(mapName string) error {
 	buildMap.Do(func() {
+		EpPolicyMap = bpf.NewMap(mapName,
```
Maybe instead of `Warning()` it should be `Fatal()`. What's the point in continuing if `fd` is 0?

qmonnet replied:
Right, it's probably useless to carry on indeed. I just changed to `Fatal()` and respined. Thanks!

qmonnet committed:
eppolicymap: exit if inner map creation fails, to avoid nil dereference

In eppolicymap.CreateWithName(), if the policy map creation in the
buildMap.Do() block fails for whatever reason, we return early and
EpPolicyMap remains nil. Calling OpenOrCreate() a few lines below
attempts to dereference that pointer to take its lock.

Ouch.

EpPolicyMap used to be created unconditionally, even if the policy map
creation would fail. This was changed in commit b384ce85bb83. We could
restore this behaviour, or we can simply exit early if inner map
creation fails, by producing a fatal error message. Let's do the latter,
as there is no point in proceeding with the creation of endpoint policy
map if creating the inner map failed.

Fixes: b384ce85bb83 ("eppolicymap: Factor out for better testing")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
